File: tests/runners/rust/tests/test_06_arrays_and_optionals.rs
// Test Case 06: Arrays and Optionals
// Tests array and optional field types

use std::io::Cursor;

use polygen_test::polygen_support::CsvRow;
use polygen_test::schema::test_collections::{ArrayTest, MixedTest, OptionalTest, Tag};
use polygen_test::schema_loaders::{BinaryIO, CsvLoadable};

fn main() {
    println!("=== Test Case 06: Arrays and Optionals ===");

    test_array_primitives();
    test_array_complex_types();
    test_optional_primitives();
    test_csv_array_parse_errors();
    test_csv_optional_parse_errors();
    test_csv_json_cell_embeds();
    test_binary_arrays_optionals();

    println!("=== All tests passed! ===");
}

fn test_array_primitives() {
    println!("  Testing ArrayTest with primitive arrays...");

    let arr = ArrayTest {
        id: 1,
        int_list: vec![1, 2, 3, 4, 5],
        string_list: vec!["one".to_string(), "two".to_string(), "three".to_string()],
        float_list: vec![1.1, 2.2, 3.3],
        bool_list: vec![true, false, true],
        tags: vec![],
    };

    assert_eq!(arr.int_list.len(), 5);
    assert_eq!(arr.int_list[0], 1);
    assert_eq!(arr.int_list[4], 5);
    assert_eq!(arr.string_list.len(), 3);
    assert_eq!(arr.string_list[1], "two");
    assert_eq!(arr.float_list.len(), 3);
    assert_eq!(arr.bool_list.len(), 3);
    assert!(arr.bool_list[0]);
    assert!(!arr.bool_list[1]);

    println!("    PASS");
}

fn test_array_complex_types() {
    println!("  Testing ArrayTest with complex type arrays...");

    let arr = ArrayTest {
        id: 2,
        int_list: vec![],
        string_list: vec![],
        float_list: vec![],
        bool_list: vec![],
        tags: vec![
            Tag { name: "Important".to_string(), color: "red".to_string() },
            Tag { name: "Review".to_string(), color: "yellow".to_string() },
        ],
    };

    assert_eq!(arr.tags.len(), 2);
    assert_eq!(arr.tags[0].name, "Important");
    assert_eq!(arr.tags[0].color, "red");
    assert_eq!(arr.tags[1].name, "Review");

    println!("    PASS");
}

fn test_optional_primitives() {
    println!("  Testing OptionalTest with optional primitives...");

    let mut opt = OptionalTest {
        id: 1,
        required_name: "Test".to_string(),
        opt_int: None,
        opt_string: None,
        opt_float: None,
        opt_bool: None,
        opt_tag: None,
    };

    assert_eq!(opt.required_name, "Test");
    assert!(opt.opt_int.is_none());
    assert!(opt.opt_string.is_none());
    assert!(opt.opt_float.is_none());
    assert!(opt.opt_bool.is_none());

    // Set values
    opt.opt_int = Some(42);
    opt.opt_string = Some("optional value".to_string());
    opt.opt_float = Some(3.14159);
    opt.opt_bool = Some(true);

    assert_eq!(opt.opt_int, Some(42));
    assert_eq!(opt.opt_string, Some("optional value".to_string()));
    assert!((opt.opt_float.unwrap() - 3.14159).abs() < 0.0001);
    assert_eq!(opt.opt_bool, Some(true));

    println!("    PASS");
}

fn test_csv_array_parse_errors() {
    println!("  Testing CSV array parse errors...");

    let bool_row = CsvRow::new(
        vec![
            "id".to_string(),
            "int_list".to_string(),
            "string_list".to_string(),
            "float_list".to_string(),
            "bool_list".to_string(),
            "tags".to_string(),
        ],
        vec![
            "1".to_string(),
            "10,20,30".to_string(),
            "one,two".to_string(),
            "1.5,2.5".to_string(),
            "yes,0,no,true".to_string(),
            "".to_string(),
        ],
    );

    let parsed = ArrayTest::from_csv_row(&bool_row).unwrap();
    assert_eq!(parsed.bool_list, vec![true, false, false, true]);

    let row = CsvRow::new(
        vec![
            "id".to_string(),
            "int_list".to_string(),
            "string_list".to_string(),
            "float_list".to_string(),
            "bool_list".to_string(),
            "tags".to_string(),
        ],
        vec![
            "1".to_string(),
            "10,not-an-int,30".to_string(),
            "one,two".to_string(),
            "1.5,2.5".to_string(),
            "true,false".to_string(),
            "".to_string(),
        ],
    );

    let err = ArrayTest::from_csv_row(&row).unwrap_err();
    assert!(format!("{}", err).contains("invalid value for int_list"));

    println!("    PASS");
}

fn test_csv_optional_parse_errors() {
    println!("  Testing CSV optional parse errors...");

    let invalid_row = CsvRow::new(
        vec![
            "id".to_string(),
            "required_name".to_string(),
            "opt_int".to_string(),
            "opt_string".to_string(),
            "opt_float".to_string(),
            "opt_bool".to_string(),
            "opt_tag".to_string(),
        ],
        vec![
            "1".to_string(),
            "Test".to_string(),
            "not-an-int".to_string(),
            "".to_string(),
            "".to_string(),
            "".to_string(),
            "".to_string(),
        ],
    );

    let err = OptionalTest::from_csv_row(&invalid_row).unwrap_err();
    assert!(format!("{}", err).contains("invalid value for opt_int"));

    let bool_row = CsvRow::new(
        vec![
            "id".to_string(),
            "required_name".to_string(),
            "opt_int".to_string(),
            "opt_string".to_string(),
            "opt_float".to_string(),
            "opt_bool".to_string(),
            "opt_tag".to_string(),
        ],
        vec![
            "2".to_string(),
            "BoolTest".to_string(),
            "".to_string(),
            "".to_string(),
            "".to_string(),
            "yes".to_string(),
            "".to_string(),
        ],
    );

    let parsed = OptionalTest::from_csv_row(&bool_row).unwrap();
    assert_eq!(parsed.opt_bool, Some(true));

    println!("    PASS");
}

fn test_csv_json_cell_embeds() {
    println!("  Testing CSV JSON-cell embedded values...");

    let array_row = CsvRow::new(
        vec![
            "id".to_string(),
            "int_list".to_string(),
            "string_list".to_string(),
            "float_list".to_string(),
            "bool_list".to_string(),
            "tags".to_string(),
        ],
        vec![
            "3".to_string(),
            "1,2".to_string(),
            "one,two".to_string(),
            "1.5,2.5".to_string(),
            "true,false".to_string(),
            r#"[{"name":"Important","color":"red"},{"name":"Review","color":"yellow"}]"#.to_string(),
        ],
    );

    let parsed_array = ArrayTest::from_csv_row(&array_row).unwrap();
    assert_eq!(parsed_array.tags.len(), 2);
    assert_eq!(parsed_array.tags[0].name, "Important");
    assert_eq!(parsed_array.tags[1].color, "yellow");

    let optional_row = CsvRow::new(
        vec![
            "id".to_string(),
            "required_name".to_string(),
            "opt_int".to_string(),
            "opt_string".to_string(),
            "opt_float".to_string(),
            "opt_bool".to_string(),
            "opt_tag".to_string(),
        ],
        vec![
            "4".to_string(),
            "WithTag".to_string(),
            "".to_string(),
            "".to_string(),
            "".to_string(),
            "".to_string(),
            r#"{"name":"Optional","color":"blue"}"#.to_string(),
        ],
    );

    let parsed_optional = OptionalTest::from_csv_row(&optional_row).unwrap();
    let opt_tag = parsed_optional.opt_tag.expect("optional tag should parse");
    assert_eq!(opt_tag.name, "Optional");
    assert_eq!(opt_tag.color, "blue");

    let mixed_row = CsvRow::new(
        vec![
            "id".to_string(),
            "opt_tags".to_string(),
            "meta".to_string(),
            "history".to_string(),
        ],
        vec![
            "5".to_string(),
            r#"[{"name":"One","color":"green"}]"#.to_string(),
            r#"{"created_by":"alice","updated_by":null,"version":3}"#.to_string(),
            r#"[{"created_by":null,"updated_by":"bob","version":1}]"#.to_string(),
        ],
    );

    let parsed_mixed = MixedTest::from_csv_row(&mixed_row).unwrap();
    assert_eq!(parsed_mixed.opt_tags.len(), 1);
    assert_eq!(parsed_mixed.opt_tags[0].name, "One");
    let meta = parsed_mixed.meta.expect("metadata should parse");
    assert_eq!(meta.created_by.as_deref(), Some("alice"));
    assert_eq!(meta.version, 3);
    assert_eq!(parsed_mixed.history.len(), 1);
    assert_eq!(parsed_mixed.history[0].updated_by.as_deref(), Some("bob"));

    let invalid_row = CsvRow::new(
        vec![
            "id".to_string(),
            "int_list".to_string(),
            "string_list".to_string(),
            "float_list".to_string(),
            "bool_list".to_string(),
            "tags".to_string(),
        ],
        vec![
            "6".to_string(),
            "".to_string(),
            "".to_string(),
            "".to_string(),
            "".to_string(),
            "not-json".to_string(),
        ],
    );
    let err = ArrayTest::from_csv_row(&invalid_row).unwrap_err();
    assert!(format!("{}", err).contains("invalid JSON for tags"));

    println!("    PASS");
}

fn test_binary_arrays_optionals() {
    println!("  Testing binary serialization with arrays and optionals...");

    let original = ArrayTest {
        id: 123,
        int_list: vec![10, 20, 30],
        string_list: vec!["a".to_string(), "b".to_string(), "c".to_string()],
        float_list: vec![1.5, 2.5],
        bool_list: vec![true, false],
        tags: vec![
            Tag { name: "Test".to_string(), color: "white".to_string() },
        ],
    };

    // Serialize
    let mut buffer = Vec::new();
    original.write_binary(&mut buffer).unwrap();

    // Deserialize
    let mut cursor = Cursor::new(&buffer);
    let loaded = ArrayTest::read_binary(&mut cursor).unwrap();

    assert_eq!(loaded.id, original.id);
    assert_eq!(loaded.int_list, original.int_list);
    assert_eq!(loaded.string_list[0], "a");
    assert!(loaded.bool_list[0]);
    assert_eq!(loaded.tags.len(), 1);
    assert_eq!(loaded.tags[0].name, "Test");

    println!("    PASS (serialized {} bytes)", buffer.len());
}

File: tests/runners/typescript/tests/test_06_arrays_and_optionals.ts
// Test Case 06: Arrays and Optionals
// Tests array and optional field types

import { mkdtempSync, rmSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import {
    TestCollections,
    loadTestCollectionsArrayTestsFromCsv,
    loadTestCollectionsArrayTestsFromJson,
    loadTestCollectionsMixedTestsFromCsv,
    loadTestCollectionsOptionalTestsFromCsv,
    fromTestCollectionsArrayTestBinary,
    fromTestCollectionsMixedTestBinary,
    fromTestCollectionsOptionalTestBinary,
    toTestCollectionsArrayTestBinary,
    toTestCollectionsMixedTestBinary,
    toTestCollectionsOptionalTestBinary,
} from '../generated/06_arrays_and_optionals/typescript/schema';

// Test ArrayTest with various array types
function testArrayTypes(): void {
    console.log("  Testing ArrayTest with various array types...");

    const arr: TestCollections.ArrayTest = {
        id: 1,
        intList: [1, 2, 3, 4, 5],
        stringList: ["a", "b", "c"],
        floatList: [1.1, 2.2, 3.3],
        boolList: [true, false, true],
        tags: [
            { name: "important", color: "red" },
            { name: "review", color: "yellow" },
        ],
    };

    console.assert(arr.id === 1, "id should be 1");
    console.assert(arr.intList.length === 5, "intList should have 5 elements");
    console.assert(arr.intList[0] === 1, "intList[0] should be 1");
    console.assert(arr.stringList.length === 3, "stringList should have 3 elements");
    console.assert(arr.tags.length === 2, "tags should have 2 elements");
    console.assert(arr.tags[0].name === "important", "tags[0].name should be important");

    console.log("    PASS");
}

// Test OptionalTest with optional fields
function testOptionalFields(): void {
    console.log("  Testing OptionalTest with optional fields...");

    // All optionals undefined
    const opt1: TestCollections.OptionalTest = {
        id: 1,
        requiredName: "Test",
        optInt: undefined,
        optString: undefined,
        optFloat: undefined,
        optBool: undefined,
        optTag: undefined,
    };

    console.assert(opt1.id === 1, "id should be 1");
    console.assert(opt1.requiredName === "Test", "requiredName should be Test");
    console.assert(opt1.optInt === undefined, "optInt should be undefined");

    // All optionals set
    const opt2: TestCollections.OptionalTest = {
        id: 2,
        requiredName: "Full",
        optInt: 42,
        optString: "hello",
        optFloat: 3.14,
        optBool: true,
        optTag: { name: "tag", color: "blue" },
    };

    console.assert(opt2.optInt === 42, "optInt should be 42");
    console.assert(opt2.optString === "hello", "optString should be hello");
    console.assert(opt2.optFloat === 3.14, "optFloat should be 3.14");
    console.assert(opt2.optBool === true, "optBool should be true");
    console.assert(opt2.optTag?.name === "tag", "optTag.name should be tag");

    console.log("    PASS");
}

// Test MixedTest with arrays and optionals
function testMixedArraysAndOptionals(): void {
    console.log("  Testing MixedTest with arrays and optionals...");

    const mixed: TestCollections.MixedTest = {
        id: 1,
        optTags: [
            { name: "tag1", color: "red" },
            { name: "tag2", color: "blue" },
        ],
        meta: {
            createdBy: "admin",
            updatedBy: undefined,
            version: 1,
        },
        history: [
            { createdBy: "user1", updatedBy: "user2", version: 1 },
            { createdBy: "user2", updatedBy: undefined, version: 2 },
        ],
    };

    console.assert(mixed.id === 1, "id should be 1");
    console.assert(mixed.optTags.length === 2, "optTags should have 2 elements");
    console.assert(mixed.meta?.createdBy === "admin", "meta.createdBy should be admin");
    console.assert(mixed.meta?.updatedBy === undefined, "meta.updatedBy should be undefined");
    console.assert(mixed.history.length === 2, "history should have 2 elements");

    console.log("    PASS");
}

function csvJsonCell(value: unknown): string {
    return `"${JSON.stringify(value).replace(/"/g, '""')}"`;
}

function testGeneratedLoaders(): void {
    console.log("  Testing generated CSV/JSON loaders...");

    const dir = mkdtempSync(join(tmpdir(), 'polygen-ts-06-'));
    try {
        const arrayCsv = join(dir, 'array_tests.csv');
        writeFileSync(
            arrayCsv,
            [
                'id,int_list,string_list,float_list,bool_list,tags',
                `1,"1,2,3","alpha,beta","1.5,2.5","yes,no,1",${csvJsonCell([
                    { name: 'important', color: 'red' },
                    { name: 'review', color: 'yellow' },
                ])}`,
            ].join('\n'),
            'utf8',
        );

        const optionalCsv = join(dir, 'optional_tests.csv');
        writeFileSync(
            optionalCsv,
            [
                'id,required_name,opt_int,opt_string,opt_float,opt_bool,opt_tag',
                `2,Full,42,hello,3.14,true,${csvJsonCell({ name: 'tag', color: 'blue' })}`,
                '3,Empty,,,,,',
            ].join('\n'),
            'utf8',
        );

        const mixedCsv = join(dir, 'mixed_tests.csv');
        writeFileSync(
            mixedCsv,
            [
                'id,opt_tags,meta,history',
                `4,${csvJsonCell([{ name: 'tag1', color: 'red' }])},${csvJsonCell({ created_by: 'admin', version: 1 })},${csvJsonCell([
                    { created_by: 'user1', updated_by: 'user2', version: 1 },
                    { created_by: 'user2', version: 2 },
                ])}`,
            ].join('\n'),
            'utf8',
        );

        const arrayJson = join(dir, 'array_tests.json');
        writeFileSync(
            arrayJson,
            JSON.stringify([
                {
                    id: 5,
                    int_list: [9, 8],
                    string_list: ['x', 'y'],
                    float_list: [1.25],
                    bool_list: [true, false],
                    tags: [{ name: 'json', color: 'green' }],
                },
            ]),
            'utf8',
        );

        const arraysFromCsv = loadTestCollectionsArrayTestsFromCsv(arrayCsv);
        console.assert(arraysFromCsv[0].intList[2] === 3, "CSV int list should parse");
        console.assert(arraysFromCsv[0].boolList[0] === true && arraysFromCsv[0].boolList[1] === false, "CSV bool list should parse yes/no");
        console.assert(arraysFromCsv[0].tags[1].color === 'yellow', "CSV JSON-cell embed list should parse");

        const optionalsFromCsv = loadTestCollectionsOptionalTestsFromCsv(optionalCsv);
        console.assert(optionalsFromCsv[0].optTag?.name === 'tag', "CSV optional embed should parse");
        console.assert(optionalsFromCsv[1].optInt === undefined, "empty optional scalar should be undefined");

        const mixedFromCsv = loadTestCollectionsMixedTestsFromCsv(mixedCsv);
        console.assert(mixedFromCsv[0].meta?.createdBy === 'admin', "CSV optional nested embed should parse");
        console.assert(mixedFromCsv[0].history[1].version === 2, "CSV nested embed list should parse");

        const arraysFromJson = loadTestCollectionsArrayTestsFromJson(arrayJson);
        console.assert(arraysFromJson[0].stringList[1] === 'y', "JSON array loader should parse snake_case fields");
        console.assert(arraysFromJson[0].tags[0].name === 'json', "JSON embed array should parse");
    } finally {
        rmSync(dir, { recursive: true, force: true });
    }

    console.log("    PASS");
}

function testBinaryRoundtrip(): void {
    console.log("  Testing generated binary roundtrip...");

    const arrayRow: TestCollections.ArrayTest = {
        id: 10,
        intList: [10, 20, 30],
        stringList: ["alpha", "beta"],
        floatList: [1.5, 2.25],
        boolList: [true, false, true],
        tags: [{ name: "binary", color: "green" }],
    };
    const arrayLoaded = fromTestCollectionsArrayTestBinary(toTestCollectionsArrayTestBinary(arrayRow));
    console.assert(arrayLoaded.id === arrayRow.id, "binary array id should roundtrip");
    console.assert(arrayLoaded.intList.join(",") === "10,20,30", "binary int list should roundtrip");
    console.assert(arrayLoaded.stringList[1] === "beta", "binary string list should roundtrip");
    console.assert(Math.abs(arrayLoaded.floatList[1] - 2.25) < 0.0001, "binary f32 list should roundtrip");
    console.assert(arrayLoaded.boolList[1] === false, "binary bool list should roundtrip");
    console.assert(arrayLoaded.tags[0].name === "binary", "binary embed list should roundtrip");

    const optionalFull: TestCollections.OptionalTest = {
        id: 11,
        requiredName: "Full",
        optInt: -7,
        optString: "present",
        optFloat: 9.5,
        optBool: true,
        optTag: { name: "optional", color: "blue" },
    };
    const optionalLoaded = fromTestCollectionsOptionalTestBinary(toTestCollectionsOptionalTestBinary(optionalFull));
    console.assert(optionalLoaded.optInt === -7, "binary optional int should roundtrip");
    console.assert(optionalLoaded.optString === "present", "binary optional string should roundtrip");
    console.assert(optionalLoaded.optFloat === 9.5, "binary optional f64 should roundtrip");
    console.assert(optionalLoaded.optBool === true, "binary optional bool should roundtrip");
    console.assert(optionalLoaded.optTag?.color === "blue", "binary optional embed should roundtrip");

    const optionalEmpty: TestCollections.OptionalTest = {
        id: 12,
        requiredName: "Empty",
        optInt: undefined,
        optString: undefined,
        optFloat: undefined,
        optBool: undefined,
        optTag: undefined,
    };
    const emptyLoaded = fromTestCollectionsOptionalTestBinary(toTestCollectionsOptionalTestBinary(optionalEmpty));
    console.assert(emptyLoaded.optInt === undefined, "empty binary optional int should stay undefined");
    console.assert(emptyLoaded.optTag === undefined, "empty binary optional embed should stay undefined");

    const mixed: TestCollections.MixedTest = {
        id: 13,
        optTags: [{ name: "history", color: "purple" }],
        meta: { createdBy: "admin", updatedBy: undefined, version: 2 },
        history: [
            { createdBy: "a", updatedBy: "b", version: 1 },
            { createdBy: undefined, updatedBy: "c", version: 2 },
        ],
    };
    const mixedLoaded = fromTestCollectionsMixedTestBinary(toTestCollectionsMixedTestBinary(mixed));
    console.assert(mixedLoaded.meta?.createdBy === "admin", "binary optional nested embed should roundtrip");
    console.assert(mixedLoaded.meta?.updatedBy === undefined, "binary nested optional field should stay undefined");
    console.assert(mixedLoaded.history[1].updatedBy === "c", "binary nested embed list should roundtrip");

    console.log("    PASS");
}

// Main
console.log("=== Test Case 06: Arrays and Optionals ===");
testArrayTypes();
testOptionalFields();
testMixedArraysAndOptionals();
testGeneratedLoaders();
testBinaryRoundtrip();
console.log("=== All tests passed! ===");
